From 236e589991bef8fe4cafd9fdef55a15c635a847d Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Wed, 09 Mar 2022 14:41:42 +0200
Subject: [PATCH] [SW-67981] greco: add polling for clearing flr int

Currently as part of FLR interrupt processing flow, FLR interrupt
is cleared at source in FLR_INT register followed by cleared in
SPI_INTR_CLR register. However SPI_INTR for FLR doesn't get cleared
as there is some latency in clearing FLR interrupt at source. To
overcome this, this patch add polling for FLR_INT to be cleared with
timeout of 10us after which interrupt is cleared in SPI_INTR_CLR
register. This makes sure the FLR interrupt is cleared at source before
clearing it in SPT_INTR register.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: Ia9ba0334759f3bd323b5720643da36a56ec4ec6e
---

diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index 9d2df9f..6987299 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -280,12 +280,23 @@
 	}
 }
 
+#define PCIE_FLR_CLR_TIMEOUT_US 10
 void pci_dev_specific_flr(void)
 {
+	u64 start;
+
 	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
 	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
 	regmap_writel(1, mmPCIE_AUX_FLR_INT);
 
+	start = timestamp_get_us();
+	while (regmap_readl(mmPCIE_AUX_FLR_INT)) {
+		if (timeout_expired(start, PCIE_FLR_CLR_TIMEOUT_US)) {
+			preboot_puts("FLR not cleared\n");
+			break;
+		}
+	}
+
 	/*
 	 * Clear PCIe SPI INTR. This is required to clear
 	 * FLR interrupt as recommended by HW team.
